DRY-up footer across email builders

benthorner opened the pull request:
https://trello.com/c/wp7WjPN6/693-draft-email-to-send-out-to-checker-subscribers-following-an-announcement

This makes a few prepatory changes in order to DRY-up
the presentation of the footer into a single presenter.
Making this change will also make it easier to track the
footer links, which may come in a future PR.

Please see the commits for more details.

kevindew commented on app/builders/digest_email_builder.rb:
```diff
@@ -34,15 +34,7 @@ def body
-      [Unsubscribe](#{unsubscribe_url})
-
-      [#{I18n.t!('emails.digests.footer_manage')}](#{PublicUrls.authenticate_url(address: subscriber.address)})
+      #{FooterPresenter.call(subscriber, subscription).strip}
```
Could the strip be done in the presenter? It feels a strange extra detail for the caller to consider.

benthorner replied:
Sure, I've amended the commit to do that. I was trying to be consistent with the presenters for content changes and messages, which inject a newline at the end. This was quite surprising, though, so I've added another commit to make them all consistent i.e. to not include a newline at the end.

kevindew reviewed (commented):
This looks great - nice work with the DigestEmailQuery class, that is a much easier read. Just a couple of small comments.

I'm a bit concerned about the use of Presenter classes as a mechanism to act as view partials but it looks like that was added long before this PR was open so I don't think that's a fair thing to try resolve here. It just adds to my concern that Presenter classes are misused most of the time.

benthorner commented on app/queries/digest_items_query.rb:
```diff
@@ -11,7 +11,15 @@ def self.call(*args)
 
   def call
-    build_results(fetch_content_changes, fetch_messages)
+    subscriptions.map { |subscription|
```
Aha! I'd read about that, and then forgot. I've amended the first commit with it 👍.

benthorner commented on app/queries/digest_items_query.rb:
```diff
@@ -20,54 +28,35 @@ def call
-      .merge(Subscription.active)
-      .order("subscriber_list_title ASC", "subscriber_list_url ASC", "messages.created_at ASC")
+      .order("subscriber_lists.title ASC", "messages.created_at ASC")
       .uniq(&:id)
```
Yes, that was confusing - that and the `order` clauses. I've added a commit for the latter, as I don't think they're necessary. I'm also confused about why we need to remove content changes for the same Content ID, as presumably they all reflect distinct changes the subscriber may want to learn about.

For messages, it's even more confusing: there's [a test](https://github.com/alphagov/email-alert-api/blob/f1687ad93e261a1ec4368072407b5221d68fcb7a/spec/queries/digest_items_query_spec.rb#L153-L166) that specifically requires that `uniq(&:id)` call to be there. Even when digests were a single email, it still doesn't make sense to only show the change under one of the lists - we're effectively doing the same for content changes with the `uniq` call there, too.

I'm concerned this might be a bit of a rabbit hole, though. While it seems like both of those `uniq` calls can be removed, I'd prefer to do that in a follow-up PR, if we do it at all.

kevindew replied:
Sure, I recall the `content_id` one being a product decision that it was confusing that you could have conflicting updates to the same content item in a digest, however I doubt that is a documented product decision and seems rather heavily caveated where things could be confusing.

I can't help but think the id one on Message is just a mistake though and that the test you'd referenced would pass without it - given ActiveRecord won't return the same model twice. Not too concerned if you want to defer it though.

---

Nice one on the ordering. I imagine prior to messages we didn't need to do any ordering in Ruby but now there are two things combined.

benthorner commented:
@kevindew glad you like the result! I was cautious about making such a sweeping change for a seemingly trivial refactoring. I think I've addressed all your comments - could you take another look?

I agree the presenters are strange, and so are the builders, which are doing more than building (by actually creating email objects). I reckon we could merge the concepts together. We could even go further and use ActionMailer views, which could lead to a more conventional use of Rails conventions. On the other hand, I don't think they're doing much harm the way they are, as long as they're consistent. Since this is an API app, there's little risk of ambiguity of purpose for the presenters.

kevindew reviewed (approved):
> I agree the presenters are strange, and so are the builders, which are doing more than building (by actually creating email objects). I reckon we could merge the concepts together. We could even go further and use ActionMailer views, which could lead to a more conventional use of Rails conventions. On the other hand, I don't think they're doing much harm the way they are, as long as they're consistent. Since this is an API app, there's little risk of ambiguity of purpose for the presenters.

Yeah I think this would all make much more sense with actual view files rather than these things being wrapped in classes. Does seem a reasonable chunk of work though and things could get rather complex trying to consolidate what should be ActionMailer what should be custom.

Ah ha I'd never underestimate the ability for different type of Presenters to end up in an API app - many of these end up with serialisers at some point.

benthorner commented:
(For some reason GitHub won't let me reply to your thread 😠.)

> I can't help but think the id one on Message is just a mistake though and that the test you'd referenced would pass without it - given ActiveRecord won't return the same model twice. Not too concerned if you want to defer it though.

So it does actually fail. The reason is that we're doing a `joins` on `matched_messages`, so we actually get the same Message multiple times. And just like for content changes, the `uniq` means it'll only appear in one digest email, even if it (arguably) should appear in all digest emails with a matching subscription, especially now they're separate.

(I do get the extra concern for content changes, although arguably we've broken away from that already with messages, since we won't dedup messages about the same action - the system simply doesn't have the data.)

I think there's a case for removing both of the `uniq`s, but this could have awkward side-effects. For example, if someone's accrued multiple near-equivalent Brexit Checker subscriptions over time, they'd start getting multiple digest emails about exactly the same things, whereas previously the messages would (probably) have all been de-duped into just one of the matching digest subscriptions i.e. only one email. Arguably we should just let that happen, and users can figure out the mess, but then it's also arguably a mess we could have helped them to avoid getting into in the first place.

Hope that makes some sense! I think it's definitely something to defer for this PR, and sufficiently subtle and tangential that I'm not sure it's worth pursuing, either - not without an actual team to run it past.

kevindew commented:
> (For some reason GitHub won't let me reply to your thread 😠.)
> 
> > I can't help but think the id one on Message is just a mistake though and that the test you'd referenced would pass without it - given ActiveRecord won't return the same model twice. Not too concerned if you want to defer it though.
> 
> So it does actually fail. The reason is that we're doing a `joins` on `matched_messages`, so we actually get the same Message multiple times. And just like for content changes, the `uniq` means it'll only appear in one digest email, even if it (arguably) should appear in all digest emails with a matching subscription, especially now they're separate.
> 
> (I do get the extra concern for content changes, although arguably we've broken away from that already with messages, since we won't dedup messages about the same action - the system simply doesn't have the data.)
> 
> I think there's a case for removing both of the `uniq`s, but this could have awkward side-effects. For example, if someone's accrued multiple near-equivalent Brexit Checker subscriptions over time, they'd start getting multiple digest emails about exactly the same things, whereas previously the messages would (probably) have all been de-duped into just one of the matching digest subscriptions i.e. only one email. Arguably we should just let that happen, and users can figure out the mess, but then it's also arguably a mess we could have helped them to avoid getting into in the first place.
> 
> Hope that makes some sense! I think it's definitely something to defer for this PR, and sufficiently subtle and tangential that I'm not sure it's worth pursuing, either - not without an actual team to run it past.

Thanks for explaining - agree on deferring.

Oh dear that's a TIL for me on the id, I thought Rails de-duped those joins. I wonder if there's something fishy about that query altogether it could be returning lots more data than is actually needed. But yes, problem for another day.
